### PR TITLE
#4541: remove unpad start at 0 limitation

### LINKED ISSUE #4541: Extend Unpad to Generalized Slice
This is needed for multi-device as well as performing a general split (@muthutt , I will get this done this week and we can use this to replace split last dim as well).

CC: @tt-aho , @davorchap

### PR DESCRIPTION
**Pull request by ntarafdar**

### CODE REVIEW

**Review by boris-drazic — APPROVED** [review]

File: tests/tt_eager/python_api_testing/unit_testing/test_unpadding_test.py [hunk 1]
```diff
@@ -9,11 +9,22 @@
...
 ):
-    inp = torch.rand(*input_tensor_shape, dtype=torch.bfloat16)
+    if input_layout == ttl.tensor.Layout.TILE:
+        two_d_shape = (input_tensor_shape[0] * input_tensor_shape[1] * input_tensor_shape[2], input_tensor_shape[3])
+        num_tiles_width = (two_d_shape[1]) // TILE_SIZE
+        num_tiles_height = (two_d_shape[0]) // TILE_SIZE
+        inp = get_debug_tensor(num_tiles_width, num_tiles_height, torch.bfloat16)
+        inp = inp.reshape(input_tensor_shape)
+    else:
+        inp = torch.rand(*input_tensor_shape, dtype=torch.bfloat16)
```
Comment (tt-aho) [inline tests/tt_eager/python_api_testing/unit_testing/test_unpadding_test.py:27]: This looks like debug code so should we revert this change?

File: tests/tt_eager/python_api_testing/unit_testing/test_unpadding_test.py [hunk 2]
```diff
@@ -56,12 +66,29 @@ def unpadding_test(
...
-    (((4, 3, 64, 64), (0, 0, 0, 0), (3, 2, 31, 31)),),
+    (
+        ((4, 3, 64, 64), (0, 0, 0, 0), (3, 2, 31, 31)),
+        ((1, 1, 64, 64), (0, 0, 0, 0), (0, 0, 31, 31)),
+        ((1, 1, 64, 64), (0, 0, 0, 0), (0, 0, 31, 63)),
+        ((1, 1, 64, 128), (0, 0, 0, 32), (0, 0, 63, 127)),
+        ((1, 1, 128, 96), (0, 0, 64, 32), (0, 0, 95, 95)),
+        ((1, 1, 128, 96), (0, 0, 64, 32), (0, 0, 95, 95)),
+        ((1, 1, 96, 64), (0, 0, 64, 32), (0, 0, 95, 63)),
+    ),
```
Comment (tt-aho) [inline tests/tt_eager/python_api_testing/unit_testing/test_unpadding_test.py:77]: Do we support unpadding at start of non-last two dims? There doesn't seem to be any test cases for this

File: tt_eager/tt_dnn/op_library/unpad/single_core/unpad_op_single_core.cpp [hunk 3]
```diff
@@ -217,11 +221,12 @@ std::pair<std::vector<uint32_t>, std::vector<uint32_t> >   get_unpad_runtime_arg
     }
 
     uint32_t num_unpadded_tiles = output_tensor.volume() / TILE_HW;
+    uint32_t start_offset = get_tiled_start_offset(input_tensor, output_tensor_start);
 
     vector<uint32_t> reader_kernel_args = {
-        input_tensor.buffer()->address(),
+        input_tensor.buffer()->address(), // + output_tensor_start[-1] * output_tensor.element_size(),
```
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/unpad/single_core/unpad_op_single_core.cpp:227]: Remove comment?

File: tt_eager/tt_dnn/op_library/unpad/unpad_op.cpp [hunk 4]
```diff
@@ -18,18 +18,41 @@ namespace tt {
...
+uint32_t get_tiled_start_offset(const Tensor &input_tensor,
+                        const Shape &output_tensor_start
+ ) {
+    uint32_t num_input_pages = input_tensor.volume() / (TILE_HW);
+    uint32_t num_pages_last_two_dim = num_input_pages / (input_tensor.shape()[0] * input_tensor.shape()[1]);
+    uint32_t num_pages_width = num_input_pages / (input_tensor.shape()[0] * input_tensor.shape()[1] * (input_tensor.shape()[2]/TILE_HEIGHT));
+    uint32_t num_pages_height = num_input_pages / (input_tensor.shape()[0] * input_tensor.shape()[1] * (input_tensor.shape()[3]/TILE_WIDTH));
+    uint32_t start_offset = ((output_tensor_start[0] * output_tensor_start[1] ) * num_pages_last_two_dim +
+                                output_tensor_start[2]/TILE_HEIGHT*num_pages_width + output_tensor_start[3]/TILE_WIDTH);
+    return start_offset;
```
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/unpad/unpad_op.cpp:30]: you should remove the hardcoding to 4D here.
Use -2 and -1 for last two dims.
Multiply all elements together except last two for where you have [0] * [1]

These don't seem to correctly take into account non-last two dim start offsets. Is this expected?

**Review by tt-aho — CHANGES_REQUESTED** [review]

File: tt_eager/tt_dnn/op_library/unpad/unpad_op.cpp (hunk 4 above)
Comment (tarafdarTT) [reply tt_eager/tt_dnn/op_library/unpad/unpad_op.cpp]: Thanks Austin, ya I changed to account for upper dims and generalized it.

File: tt_eager/tt_dnn/op_library/unpad/unpad_op.cpp [hunk 5]
```diff
@@ -19,17 +19,80 @@ namespace tt {
...
+}
+
+uint32_t get_rm_start_offset(
+                        const Tensor &tensor,
+                        const Shape &output_tensor_start
+ ) {
+
+    uint32_t num_pages = tensor.volume() / tensor.shape()[-1];
+    uint32_t upper_dims_compressed = get_upper_dims_compressed(tensor);
+    uint32_t num_pages_width = num_pages / (upper_dims_compressed * (tensor.shape()[-2]));
```
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/unpad/unpad_op.cpp:81]: Isnt this value always just 1?

Comment (tarafdarTT) [reply tt_eager/tt_dnn/op_library/unpad/unpad_op.cpp:81]: yea it is , haha its copy paste from the code in tiled where it can be multiple page widths. I'll change that.

File: tt_eager/tt_dnn/op_library/unpad/unpad_op.cpp [hunk 6]
```diff
@@ -19,17 +19,80 @@ namespace tt {
...
+    }
+    else {
+        uint32_t page_width = tensor.shape()[-1];
+        num_pages /= page_width;
+    }
+    // offset for every dim except last 2
+    uint32_t start_offset = 0;
+    for(int dim_outer=0; dim_outer<tensor.shape().rank() - 2; dim_outer++) {
+        uint32_t compressed_dims = 1;
+        for(int dim_inner = dim_outer; dim_inner >= 0; dim_inner--) {
```
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/unpad/unpad_op.cpp:45]: Can we use unsigned ints for consistency in the loop vars since rank() returns an unsigned type?

**Review by tt-aho — APPROVED** [review]
lgtm.

Also need to update the unpad docstring in `tt_eager/tt_lib/csrc/tt_lib_bindings_tensor_dm_ops.cpp` since it currently specifies all zeros

File: tt_eager/tt_dnn/op_library/unpad/unpad_op.cpp (hunk 6 above)
Comment (tarafdarTT) [reply tt_eager/tt_dnn/op_library/unpad/unpad_op.cpp]: that makes sense, just did that.

**Comment by tt-aho** [comment]
Just realized there is a minor issue with the RM version if you had an actual 1d tensor input, since your indexing/helper functions being used assumes rank of at least 2. Maybe make `get_upper_start_offset` take an arg that is 2 for tile, 1 for rm?

**Comment by tarafdarTT** [comment]
> Just realized there is a minor issue with the RM version if you had an actual 1d tensor input, since your indexing/helper functions being used assumes rank of at least 2. Maybe make `get_upper_start_offset` take an arg that is 2 for tile, 1 for rm?

I added an if guard around tensors that are less than 2 in rank.
It would skip the loop for 1d tensor (returning the default value of 0 for offset, and 1 for upper dims compressed).

**Comment by tt-aho** [comment]
> > Just realized there is a minor issue with the RM version if you had an actual 1d tensor input, since your indexing/helper functions being used assumes rank of at least 2. Maybe make `get_upper_start_offset` take an arg that is 2 for tile, 1 for rm?
> 
> I added an if guard around tensors that are less than 2 in rank. It would skip the loop for 1d tensor (returning the default value of 0 for offset, and 1 for upper dims compressed).

Should we just guard the entire body of `get_rm_start_offset` to return 0 instead?
Otherwise it's just doing unused computations and has a [-2] index in the function body.

**Comment by tarafdarTT** [comment]
> > > Just realized there is a minor issue with the RM version if you had an actual 1d tensor input, since your indexing/helper functions being used assumes rank of at least 2. Maybe make `get_upper_start_offset` take an arg that is 2 for tile, 1 for rm?
> > 
> > 
> > I added an if guard around tensors that are less than 2 in rank. It would skip the loop for 1d tensor (returning the default value of 0 for offset, and 1 for upper dims compressed).
> 
> Should we just guard the entire body of `get_rm_start_offset` to return 0 instead? Otherwise it's just doing unused computations and has a [-2] index in the function body.

True!! Ya I moved the guard to around all compute in both helper functions :)
